src/plugins/data_source/server/client/data_source_client.ts

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Client } from '@opensearch-project/opensearch';
+import {
+  Logger,
+  OpenSearchClient,
+  SavedObject,
+  SavedObjectsClientContract,
+  SavedObjectsErrorHelpers,
+} from '../../../../../src/core/server';
+import { CREDENTIAL_SAVED_OBJECT_TYPE, DATA_SOURCE_SAVED_OBJECT_TYPE } from '../../common';
+import {
+  CredentialMaterials,
+  CredentialSavedObjectAttributes,
+} from '../../common/credentials/types';
+import { DataSourceAttributes } from '../../common/data_sources';
+import { DataSourcePluginConfigType } from '../../config';
+import { DataSourceServiceSetup } from '../data_source_service';
+
+/**
+ * Represents an OpenSearch cluster API client created by the platform.
+ * It allows to call API on behalf of the user(credential) associated to "data source"
+ *
+ * @public
+ **/
+export interface IDataSourceClient {
+  /**
+   * Creates a {@link OpenSearchClient } bound to given data source
+   */
+  asDataSource: (dataSourceId: string) => Promise<OpenSearchClient>;
+}
+
+interface DataSourceClientCtorParams {
+  dataSourceService: DataSourceServiceSetup;
+  logger: Logger;
+  scopedSavedObjectsClient: SavedObjectsClientContract;
+  config: DataSourcePluginConfigType;
+}
+export class DataSourceClient implements IDataSourceClient {
+  private dataSourceService: DataSourceServiceSetup;
+  private log: Logger;
+  // scoped saved object client to fetch save object on behalf of user
+  private scopedSavedObjectClient: SavedObjectsClientContract;
+  private config: DataSourcePluginConfigType;
+
+  constructor(ctorParams: DataSourceClientCtorParams) {
+    this.dataSourceService = ctorParams.dataSourceService;
+    this.log = ctorParams.logger;
+    this.scopedSavedObjectClient = ctorParams.scopedSavedObjectsClient;
+    this.config = ctorParams.config;
+  }
+
+  async asDataSource(dataSourceId: string) {
+    const dataSource = await this.getDataSource(dataSourceId);
+    const rootClient = this.getRootClient(dataSource.attributes, this.config);
+    const credential = await this.getCredential(dataSource.references[0].id); // assuming there is 1 and only 1 credential for each data source
+
+    return this.getQueryClient(rootClient, credential.attributes, dataSource.attributes.withAuth);
+  }
+
+  private async getDataSource(dataSourceId: string): Promise<SavedObject<DataSourceAttributes>> {
+    try {
+      const dataSource = await this.scopedSavedObjectClient.get<DataSourceAttributes>(
+        DATA_SOURCE_SAVED_OBJECT_TYPE,
+        dataSourceId
+      );
+      return dataSource;
+    } catch (error: any) {
+      // it will cause 500 error when failed to get saved objects, need to handle such error gracefully
+      throw SavedObjectsErrorHelpers.createBadRequestError(error.message);
```

Comment:
add some log

Reply:
have done some refactor, log is added at the upper level,